From d27b96465848f7fb515fea195accbcf8d2d1248d Mon Sep 17 00:00:00 2001
From: "K, Gururaj" <gururaj.k@intel.com>
Date: Tue, 17 Apr 2018 09:52:05 +0530
Subject: [PATCH 07/11] usb: gadget : f_fs: Handle endpoint shutdown while
 switching between gadget modes randomly

Switching between gadget modes randomly may cause kernel panic due abrupt endpoint shutdown.
Handle endpoint shutdown to prevent kernel panic

Change-Id: Ic90e0fbcc1b07dfeb049d7c7938ee520fce4b035
Jira: None
Test: None
Signed-off-by: K, Gururaj <gururaj.k@intel.com>
---
 drivers/usb/gadget/function/f_fs.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/usb/gadget/function/f_fs.c b/drivers/usb/gadget/function/f_fs.c
index 1746754..054e330 100644
--- a/drivers/usb/gadget/function/f_fs.c
+++ b/drivers/usb/gadget/function/f_fs.c
@@ -1015,6 +1015,10 @@ static ssize_t ffs_epfile_io(struct file *file, struct ffs_io_data *io_data)
 
 		if (interrupted)
 			ret = -EINTR;
+		else  if (epfile->ep != ep) {
+    /* In the meantime, endpoint got disabled or changed. */
+		ret = -ESHUTDOWN;
+		}
 		else if (io_data->read && ep->status > 0)
 			ret = __ffs_epfile_read_data(epfile, data, ep->status,
 						     &io_data->data);
-- 
1.9.1

